services/spar/src/Spar/App.hs

```diff
@@ -166,22 +195,35 @@ instance SPStoreIdP SparError Spar where
 
 -- | 'wrapMonadClient' with an 'Env' in a 'ReaderT', and exceptions. If you
 -- don't need either of those, 'wrapMonadClient' will suffice.
-wrapMonadClientWithEnv :: forall a. ReaderT Data.Env (ExceptT TTLError Cas.Client) a -> Spar a
+wrapMonadClientWithEnv :: forall r a. ReaderT Data.Env (ExceptT TTLError Cas.Client) a -> Spar r a
 wrapMonadClientWithEnv action = do
   denv <- Data.mkEnv <$> (sparCtxOpts <$> ask) <*> (fromTime <$> getNow)
   either (throwSpar . SparCassandraTTLError) pure =<< wrapMonadClient (runExceptT $ action `runReaderT` denv)
 
+instance Member (Final IO) r => Catch.MonadThrow (Sem r) where
+  throwM = embedFinal . Catch.throwM @IO
+
+instance Member (Final IO) r => Catch.MonadCatch (Sem r) where
+  catch m handler = withStrategicToFinal @IO $ do
+    m' <- runS m
+    st <- getInitialStateS
+    handler' <- bindS handler
+    pure $ m' `Catch.catch` \e -> handler' $ e <$ st
```

Comment:
I'm not familiar with strategies, so I don't get the combination of `runS`, `getInitialStateS` and `bindS`. @fisx , do you understand this part?

Reply:
I just found this: https://reasonablypolymorphic.com/blog/tactics/

I don't think it's very important, though: this is something very unlikely to get any bugs past the type system *and* the tests, and the instance will go away as we make more progress with refactoring.  Then again I suspect understanding tactical and strategy may help us understand later versions of this as well.  :)

Reply:
These are low-level poslysemy constructs. Inside `withStrategicToFinal`, `runS` spiritually turns `Sem r a` into `IO a`, by running all *future* interpreters on it. Bind is the same, but for kleisli arrows.

The `getInitialStateS` lets us look at all of the accumulated functorial state up to this point in the program, and manipulate it as we please. For example, if there is a `State s` effect still to be interpreteted, we need to keep track of what state it has!

But @fisx is correct in that we'll very clearly see problems in CI if they are here.